### PR TITLE
Write-in select option support

### PR DESCRIPTION
**Pull request by solocommand**
Adds support for write-in values on custom select options.
Also resolves bug where a selected value couldn't be unset.

## Field definition
image

```graphql
mutation UpdateSelectField($input: UpdateSelectFieldMutationInput!) {
  updateSelectField(input: {
    id: "618ab74dcd3e2f0147386c42",
    name: "Game preference",
    label: "What is your favorite game?",
    externalId: {
      identifier: { type: "thud", value: "qux" },
      namespace: { provider: "foo", tenant: "bar", type: "baz" }
    },
    options: [
      {
        id: "618ab74dcd3e2f0147386c43",
        label: "Scrabble",
        externalIdentifier: "foo-bar",
        canWriteIn: false
      },
      {
        id: "626ab183cd3af80041f3cbc7",
        label: "Other",
        externalIdentifier: "foo-other",
        canWriteIn: true
      }
    ]
  }) {
    id
    type
    name
    label
    required
    active
    options {
      id
      label
      externalIdentifier
      canWriteIn
    }
    externalId {
      id
       identifier {
        type
        value
      }
      namespace {
        provider
        tenant
        type
      }
    }
    createdAt
    updatedAt
  }
}
```

## User answers
image

image

image

```graphql
mutation UpdateAppUserCustomSelectAnswers($input: UpdateAppUserCustomSelectAnswersMutationInput!) {
  updateAppUserCustomSelectAnswers(input: {
    id: "some-user-id",
    answers: [
      {
        fieldId: "618ab74dcd3e2f0147386c42",
        optionIds: ["626ab183cd3af80041f3cbc7"],
        writeInValues: [
          {
            optionId: "626ab183cd3af80041f3cbc7",
            value: "foo-bar-baz-qux-quzbot"
          }
        ]
      }
    ]
  }) {
    id
    email
    domain
    name
    customSelectFieldAnswers(input: { onlyActive: true }) {
      id
      field {
        id
        name
        label
      }
      answers {
        id
        label
        writeInValue
      }
    }
  }
}
```

### CODE REVIEW

**Comment by solocommand** [comment]
[Write-in support for select fields](https://trello.com/c/upRaKFsC/2-write-in-support-for-select-fields)

File: services/graphql/src/graphql/definitions/field.js [hunk 1]
```diff
@@ -151,6 +151,8 @@ type SelectFieldOption {
   label: String!
   "The external identifier value for this option. Only used when an external ID + namespace is associated with this field."
   externalIdentifier: String
+  "Whether free-form, write-in values are supported."
+  canWriteIn: Boolean!
```
Comment (zarathustra323) [inline services/graphql/src/graphql/definitions/field.js]: Allow this to be nullable

File: services/graphql/src/graphql/definitions/app-user.js [hunk 2]
```diff
@@ -146,7 +146,21 @@ type AppUserCustomSelectFieldAnswer {
   "Whether the user has answered the question."
   hasAnswered: Boolean!
   "The answered field option(s). This will always be an array, even if the field is a single-select only. An empty value signifies a non, or no longer valid, answer. It's up to the implementing components to account for this."
-  answers: [SelectFieldOption!]!
+  answers: [SelectFieldOptionAnswer!]!
+}
+
+type SelectFieldOptionAnswer {
+  "BC support: The select option ID. Also used as the option value."
```
Comment (zarathustra323) [inline services/graphql/src/graphql/definitions/app-user.js]: Remove the BC comments

File: services/application/src/actions/user/update-custom-select-answers.js [hunk 3]
```diff
@@ -22,24 +22,28 @@ module.exports = async ({
...
-  const userObj = user.customSelectFieldAnswers.reduce(
-    (obj, item) => ({ ...obj, [item._id]: item.values }), {},
-  );
+  const userObj = user.customSelectFieldAnswers.reduce((obj, { _id, values, writeInValues }) => ({
+    ...obj,
+    [_id]: { values, writeInValues },
+  }), {});
 
   const newAnswers = answers
-    .filter(({ optionIds }) => optionIds.length) // ignore/unset fields without options
```
Comment (zarathustra323) [inline services/application/src/actions/user/update-custom-select-answers.js:30]: Put this back for now so that inactive answers aren't removed

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Review by zarathustra323 — APPROVED** [review]

**Comment by solocommand** [comment]
[Support write-in values on select fields](https://trello.com/c/GB789yV5/3-support-write-in-values-on-select-fields)

**Pull request merged**